decoder.go

```diff
@@ -631,10 +635,19 @@ func (d *decoder) decodeStruct(name string, node ast.Node, result reflect.Value)
 		}
 	}
 
-	usedKeys := make(map[string]struct{})
 	decodedFields := make([]string, 0, len(fields))
 	decodedFieldsVal := make([]reflect.Value, 0)
 	unusedKeysVal := make([]reflect.Value, 0)
+
+	// fill unusedNodeKeys with keys from the AST
+	// a slice because we have to do equals case fold to match Filter
+	unusedNodeKeys := make([]string, 0)
+	for _, item := range list.Items {
+		for _, k := range item.Keys {
+			unusedNodeKeys = append(unusedNodeKeys, k.Token.Value().(string))
```

Comment:
Is it safe to type assert `k.Token.Value()` to a `string`? The [docs](https://godoc.org/github.com/hashicorp/hcl/hcl/ast#ObjectKey) say it can be a string _or an identifier_. Is there a test exercising that this type assertion safe if the key is an identifier?

Reply:
That's a great question, I was following the existing code in this method. Looking at the token code it's not always safe but I should be able to check before casting.

Reply:
for Keys, identifiers and strings are both strings, and we do want to call value because it may manipulate the string value